File: leetcode/algorithms/58_length_of_last_word/main.py
class LengthOfLastWord:
    """
    # Built-in function
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def lengthOfLastWord(self, s: str) -> int:
        return len(list(filter(lambda s_: s_.strip() != "", s.split(" "))).pop(-1))


    # Solution
    """
    # Solution 1
    # 
    # Reverse Traversal
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def solution1(self, s: str) -> int:
        length = 0
        i = len(s) - 1
        
        while i >= 0 and s[i] == ' ':
            i -= 1
            
        while i >= 0 and s[i] != ' ':
            length += 1
            i -= 1
            
        return length

    """
    # Solution 2
    # 
    # Built-in function
    # 
    # Complexities:
    #   N - Length of `s`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution2(self, s: str) -> int:
        words = s.split()
        
        return len(words[-1])

File: leetcode/algorithms/58_length_of_last_word/main.c
#include <string.h>

/**
 * Reverse Traversal
 * 
 * Complexities:
 *   N - Length of `s`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
int lengthOfLastWord(char* s) {
    int i = 0;
    while (s[i] != '\0') {
        i++;
    }

    int result = 0;
    for (; i >= 0; i--) {
        if (s[i] != ' ' && s[i] != '\0') {
            result++;
        } else if (s[i] = ' ' && result > 0) {
            break;
        }
    }

    return result;
}


// Solution
/**
 * Reverse Traversal
 * 
 * Complexities:
 *   N - Length of `s`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
int solution(char* s) {
    int len = strlen(s);
    int length = 0;
    int i = len - 1;

    while (i >= 0 && s[i] == ' ') {
        i--;
    }

    while (i >= 0 && s[i] != ' ') {
        length++;
        i--;
    }

    return length;
}
